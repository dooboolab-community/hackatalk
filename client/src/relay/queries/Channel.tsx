import {graphql} from 'react-relay';

export const channelsQuery = graphql`
  query ChannelsQuery($first: Int!, $after: String, $withMessage: Boolean) {
    ...MainChannelComponent_channel
      @arguments(first: $first, after: $after, withMessage: $withMessage)
  }
`;

export const findOrCreatePrivateChannel = graphql`
  mutation ChannelFindOrCreatePrivateChannelMutation($peerUserIds: [String!]!) {
    findOrCreatePrivateChannel(peerUserIds: $peerUserIds) {
      id
      name
      channelType
    }
  }
`;
