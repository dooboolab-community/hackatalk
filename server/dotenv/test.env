JWT_SECRET=dooboolab
DATABASE_URL="postgresql://test:test!@localhost:5432/postgres?schema=test"
