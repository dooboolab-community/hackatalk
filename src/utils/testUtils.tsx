import React, { ReactElement } from 'react';

import RootProvider from '../providers';
import { ThemeType } from '../types';

export const createTestElement = (
  child: ReactElement,
  themeType?: ThemeType,
): ReactElement => (
  <RootProvider initialThemeType={themeType}>{child}</RootProvider>
);

export const createTestProps = (
  obj: object = {},
  moreScreenProps?: object,
): object | unknown | any => ({
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
    resetRoot: jest.fn(),
  },
  screenProps: {
    changeThemeType: jest.fn(),
    ...moreScreenProps,
  },
  ...obj,
});
