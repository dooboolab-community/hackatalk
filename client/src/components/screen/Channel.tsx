import * as Notifications from 'expo-notifications';

import type {
  ChannelsQuery,
  ChannelsQueryResponse,
  ChannelsQueryVariables,
} from '../../__generated__/ChannelsQuery.graphql';
import { FlatList, Platform, TouchableOpacity, View } from 'react-native';
import React, { FC, Suspense, useEffect, useMemo, useState } from 'react';
import {
  graphql,
  useLazyLoadQuery,
  usePaginationFragment,
  useQueryLoader,
} from 'react-relay/hooks';

import { AdMobBanner } from 'expo-ads-admob';
import { Channel } from '../../types/graphql';
import type {
  ChannelComponent_channel$key,
} from '../../__generated__/ChannelComponent_channel.graphql';
import { ChannelLastMessageQuery } from '../../__generated__/ChannelLastMessageQuery.graphql';
import ChannelListItem from '../shared/ChannelListItem';
import EmptyListItem from '../shared/EmptyListItem';
import { LoadingIndicator } from 'dooboo-ui';
import { MainStackNavigationProps } from '../navigation/MainStackNavigator';
import { SvgPlus } from '../../utils/Icons';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useNavigation } from '@react-navigation/native';
import { useThemeContext } from '@dooboo-ui/theme';

const Container = styled.View`
  flex: 1;
  background: ${({ theme }): string => theme.background};
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Fab = styled.View`
  width: 56px;
  height: 56px;
  border-radius: 28px;
  justify-content: center;
  align-items: center;
  background: ${({ theme }): string => theme.fab};
`;

const ITEM_CNT = 20;

const channelsQuery = graphql`
  query ChannelsQuery(
    $first: Int
    $last: Int
    $before: String
    $after: String
    $withMessage: Boolean
  ) {
    ...ChannelComponent_channel @arguments(
      first: $first
      last: $last
      before: $before
      after: $after
      withMessage: $withMessage
    )
  }
`;

const channelsPaginationFragment = graphql`
  fragment ChannelComponent_channel on Query
    @argumentDefinitions(
      first: {type: "Int"}
      last: {type: "Int"}
      before: {type: "String"}
      after: {type: "String"}
      withMessage: {type: "Boolean"}
    )
    @refetchable(queryName: "Channels") {
      channels(
        first: $first
        last: $last
        before: $before
        after: $after
        withMessage: $withMessage
      )
      @connection(key: "ChannelComponent_channels" filters: ["withMessage"]) {
        edges {
          cursor
          node {
            id
            channelType
            name
            memberships(excludeMe: true) {
              user {
                name
                nickname
                thumbURL
                photoURL
              }
            }
            lastMessage {
              id
              messageType
              text
              imageUrls
              fileUrls
              createdAt
            }
          }
        }
        pageInfo {
          hasNextPage
          endCursor
        }
      }
    }
`;

const lastMessageQuery = graphql`
  query ChannelLastMessageQuery($messageId: String!) {
    message(id: $messageId) {
      id
      messageType
      text
      imageUrls
      fileUrls
      createdAt
      sender {
        id
      }
      channel {
        id
        lastMessage {
          id
        }
      }
    }
  }
`;

interface ChannelProps {
  channel: ChannelComponent_channel$key;
  searchArgs: ChannelsQueryVariables;
}

const ChannelsFragment: FC<ChannelProps> = ({
  channel,
  searchArgs,
}) => {
  const navigation = useNavigation();

  const {
    data,
    loadNext,
    isLoadingNext,
    refetch,
  } = usePaginationFragment<ChannelsQuery, ChannelComponent_channel$key>(
    channelsPaginationFragment,
    channel,
  );

  const [, loadLastMessage] = useQueryLoader<ChannelLastMessageQuery>(lastMessageQuery);
  const [bannerError, setBannerError] = useState<boolean>(false);

  useEffect(() => {
    const subscription = Notifications.addNotificationResponseReceivedListener(async (response) => {
      const messageId = JSON.parse(response.notification.request.content.data.data as string).messageId;

      if (typeof messageId === 'string') {
        loadLastMessage({ messageId });
      }
    });

    // Add notification handler.
    const responseListener = Notifications.addNotificationReceivedListener((event) => {
      const messageId = JSON.parse(event.request.content.data.data as string).messageId;

      loadNext(ITEM_CNT);

      if (typeof messageId === 'string') {
        loadLastMessage({ messageId });
      }
    });

    // Clean up : remove notification handler.
    return () => {
      Notifications.removeNotificationSubscription(responseListener);
      subscription.remove();
    };
  }, []);

  const onEndReached = (): void => {
    loadNext(ITEM_CNT);
  };

  const renderItem = ({
    item,
    index,
  }: {
    item: { node: Channel, cursor: string };
    index: number;
  }): React.ReactElement => {
    if (!item.node?.memberships || !item.node?.memberships.length) return <View key={index}/>;

    return (
      <ChannelListItem
        testID={`list-item-${index}`}
        key={index.toString()}
        item={item.node}
        onPress={(): void => {
          navigation.navigate('Message', {
            channel: item.node,
            users: item.node?.memberships?.map((membership) => membership?.user),
          });
        }}
      />
    );
  };

  const channels = useMemo(() => {
    return data?.channels?.edges ?? [];
  }, [data?.channels?.edges]);

  return <FlatList
    style={{
      alignSelf: 'stretch',
    }}
    contentContainerStyle={
      channels.length === 0
        ? {
          flex: 1,
          alignSelf: 'stretch',
          alignItems: 'center',
          justifyContent: 'center',
        }
        : null
    }
    keyExtractor={(_, index): string => index.toString()}
    // @ts-ignore
    data={channels}
    renderItem={renderItem}
    ListHeaderComponent={
      !bannerError
        ? Platform.select({
          android: <AdMobBanner
            bannerSize="fullBanner"
            // adUnitID="ca-app-pub-3940256099942544/6300978111"
            adUnitID="ca-app-pub-7837089095803162/8109702961"
            onDidFailToReceiveAdWithError={() => setBannerError(true)}
          />,
          ios: <AdMobBanner
            bannerSize="fullBanner"
            // adUnitID="ca-app-pub-3940256099942544/2934735716"
            adUnitID="ca-app-pub-7837089095803162/5084068464"
            onDidFailToReceiveAdWithError={() => setBannerError(true)}
          />,
        })
        : null
    }
    ListEmptyComponent={
      <EmptyListItem>{getString('NO_CHANNELLIST')}</EmptyListItem>
    }
    ListFooterComponent={
      <View style={{ height: 60 }} />
    }
    refreshing={isLoadingNext}
    onRefresh={() => {
      refetch(searchArgs, { fetchPolicy: 'network-only' });
    }}
    onEndReachedThreshold={0.1}
    onEndReached={onEndReached}
  />;
};

interface ContentProps {
  searchArgs: ChannelsQueryVariables;
}

const ContentContainer: FC<ContentProps> = ({
  searchArgs,
}) => {
  const data: ChannelsQueryResponse =
    useLazyLoadQuery<ChannelsQuery>(
      channelsQuery,
      searchArgs,
      { fetchPolicy: 'store-or-network' },
    );

  return <ChannelsFragment
    channel={data}
    searchArgs={searchArgs}
  />;
};

interface Props {
  navigation: MainStackNavigationProps<'Message'>;
}

const Screen: FC<Props> = () => {
  const { theme } = useThemeContext();
  const navigation = useNavigation();

  const searchArgs: ChannelsQueryVariables = {
    first: ITEM_CNT,
    withMessage: true,
  };

  return (
    <Container>
      <Suspense fallback={<LoadingIndicator/>}>
        <ContentContainer searchArgs={searchArgs}/>
      </Suspense>
      <TouchableOpacity
        activeOpacity={0.65}
        style={{
          position: 'absolute',
          right: 20,
          bottom: 44,
        }}
        onPress={(): void => navigation.navigate('ChannelCreate')}
      >
        <Fab>
          <SvgPlus fill={theme.background}/>
        </Fab>
      </TouchableOpacity>
    </Container>
  );
};

export default Screen;
