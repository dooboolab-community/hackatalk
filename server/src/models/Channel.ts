import { booleanArg, objectType } from 'nexus';

import { assert } from '../utils/assert';
import { relayToPrismaPagination } from '../utils/pagination';

export const Channel = objectType({
  name: 'Channel',
  definition(t) {
    t.model.id();
    t.model.channelType();
    t.model.name();
    t.model.lastMessageId();
    t.model.createdAt();
    t.model.updatedAt();
    t.model.deletedAt();

    t.field('lastMessage', {
      type: 'Message',
      description: 'Get latest message sent to the channel.',

      resolve: async ({ id }, args, ctx) => {
        const message = await ctx.prisma.message.findFirst({
          where: {
            channel: { id },
          },
          include: {
            sender: true,
          },
          orderBy: { createdAt: 'desc' },
        });

        return message;
      },
    });

    t.connectionField('messages', {
      type: 'Message',

      nodes: async ({ id }, args, { prisma, userId }) => {
        assert(userId, 'Not authorized.');

        const { after, before, first, last } = args;

        const blockedUsers = await prisma.blockedUser.findMany({
          select: { blockedUserId: true },
          where: { userId },
        });

        const blockedUsersInArray = blockedUsers.map(
          (user) => user.blockedUserId,
        );

        return prisma.message.findMany({
          ...relayToPrismaPagination({
            after,
            before,
            first,
            last,
          }),
          where: {
            channel: { id },
            senderId: {
              notIn: blockedUsersInArray,
            },
          },
          include: { sender: true },
        });
      },
    });

    t.list.nonNull.field('memberships', {
      type: 'Membership',
      description:
        'Get memberships assigned to channel. If excludeMe is set, it will not return authenticated user.',
      args: { excludeMe: booleanArg() },

      resolve: ({ id }, { excludeMe }, { prisma, userId }) => {
        if (excludeMe)
          return prisma.membership.findMany({
            where: {
              channel: { id },
              user: {
                id: { not: userId ?? undefined },
              },
            },
            include: { user: true },
          });

        return prisma.membership.findMany({
          where: { channel: { id } },
          include: { user: true },
        });
      },
    });
  },
});
