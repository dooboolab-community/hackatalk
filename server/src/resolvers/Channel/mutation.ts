import {
  ChannelType,
  MessageType,
} from '@prisma/client';
import {
  changeVisibilityWhenInvisible,
  createMemberships,
  createNewChannel,
  findChannelWithUserIds,
  findExistingChannel,
  findPrivateChannelWithUserIds,
} from '../../services/ChannelService';
import { inputObjectType, list, mutationField, nonNull, stringArg } from '@nexus/schema';

import { getUserId } from '../../utils/auth';

export const MessageCreateInput = inputObjectType({
  name: 'MessageCreateInput',
  definition(t) {
    t.messageType('messageType');
    t.string('text');
    t.list.string('imageUrls');
    t.list.string('fileUrls');
  },
});

export const ChannelCreateInput = inputObjectType({
  name: 'ChannelCreateInput',
  definition(t) {
    t.channelType('channelType');
    t.string('name');
    t.list.string('userIds');
  },
});

export const createChannel = mutationField('createChannel', {
  type: 'Channel',

  args: {
    channel: ChannelCreateInput,
    message: MessageCreateInput,
  },

  description: `Creates channel of [ChannelType].
  The private channel is unique by the unique members while
  the public channel can be created by each request.
  The public channel is something like an open chat while
  private channel is all kinds of direct messages.
  The [Membership] of private channel will be identical to all users which is (member).
  
  <Optional> The channel can be created with message of [MessageType].
  Please becareful when creating message and provide proper [MessageType].
  This query will return [Channel] with [Membership] without [Message] that has just created.
  `,

  resolve: async (_, { channel, message }, ctx) => {
    const userId = getUserId(ctx);
    const { channelType = ChannelType.private, userIds = [], name } = channel;

    interface Message {
      text?: string;
      messageType?: MessageType,
      fileUrls?: string[],
      imageUrls?: string[],
    }

    const createMessage = (
      { text, messageType = MessageType.text, fileUrls = [], imageUrls = [] } : Message,
      channelId: string,
    ) => ctx.prisma.message.create({
      data: {
        text,
        messageType,
        fileUrls: { set: fileUrls },
        imageUrls: { set: imageUrls },
        channel: { connect: { id: channelId } },
        sender: { connect: { id: userId } },
      },
    });

    const isPrivateChannel = channelType === ChannelType.private;

    if (isPrivateChannel) {
      const hasNoMembersToChat = userIds.length === 0;

      if (hasNoMembersToChat) {
        throw new Error('User has no members to chat within private channel.');
      }

      const existingChannel = await findChannelWithUserIds([userId, ...userIds]);

      if (existingChannel) {
        changeVisibilityWhenInvisible(userId, existingChannel);
        await createMemberships(existingChannel.id, userIds);
        message && (await createMessage(message, existingChannel.id));

        return existingChannel;
      }
    }

    const { id } = await createNewChannel(isPrivateChannel, userId, name);

    await createMemberships(id, userIds);
    message && (await createMessage(message, id));

    const getChannel = (channelId: string) => ctx.prisma.channel.findUnique({
      where: { id: channelId },
      include: {
        membership: true,
      },
    });

    return getChannel(id);
  },
});

export const findOrCreatePrivateChannel = mutationField('findOrCreatePrivateChannel', {
  type: 'Channel',
  args: { peerUserIds: nonNull(list(stringArg())) },
  description: 'Find or create channel associated to peer user id.',

  resolve: async (parent, { peerUserIds }, ctx) => {
    const userId = getUserId(ctx);
    const existingChannel = await findPrivateChannelWithUserIds([userId, ...peerUserIds]);

    if (existingChannel) {
      peerUserIds.forEach((id: string) => changeVisibilityWhenInvisible(id, existingChannel));

      return existingChannel;
    }

    const channel = await createNewChannel(true, userId);

    await createMemberships(channel.id, peerUserIds);

    return channel;
  },
});

export const leaveChannel = mutationField('leaveChannel', {
  type: 'Membership',
  args: { channelId: nonNull(stringArg()) },

  description: `User leaves [public] channel.
  Users cannot leave the [private] channel
  and rather this is going to be invisible in [Membership].
  This will reset to true when new [Message] is created to channel.
  User will leave the [public] channel and membership will be removed.
  `,

  resolve: async (parent, { channelId }, ctx) => {
    const userId = getUserId(ctx);
    const channel = await findExistingChannel(channelId);

    if (channel.channelType === 'public') {
      return ctx.prisma.membership.delete({
        where: {
          userId_channelId: {
            userId,
            channelId,
          },
        },
      });
    }

    return ctx.prisma.membership.update({
      where: {
        userId_channelId: {
          channelId,
          userId,
        },
      },
      data: {
        isVisible: false,
      },
    });
  },
});

export const inviteUsersToChannel = mutationField('inviteUsersToChannel', {
  type: 'Channel',
  args: { channelId: nonNull(stringArg()), userIds: nonNull(list(stringArg())) },
  description: 'Adds some users into [public] channel.',

  resolve: async (_, { channelId, userIds }, ctx) => {
    const userId = getUserId(ctx);
    const channel = await findExistingChannel(channelId);

    if (channel.channelType === 'private') {
      throw new Error("You can't add some users to private channel.");
    }

    const membership = await ctx.prisma.membership.findFirst({
      where: {
        userId,
        channelId: channel.id,
      },
    });

    const hasAdminPermission = membership.membershipType !== 'admin' && membership.membershipType !== 'owner';

    if (hasAdminPermission) {
      throw new Error('You should have admin or owner membership to add some users to the channel.');
    }

    for (const userId of userIds) {
      const membership = await ctx.prisma.membership.findFirst({
        where: {
          channelId,
          userId,
        },
      });

      if (!membership) {
        await ctx.prisma.membership.create({
          data: {
            user: {
              connect: { id: userId },
            },
            channel: {
              connect: { id: channelId },
            },
          },
        });
      }
    }

    return channel;
  },
});

export const kickUsersFromChannel = mutationField('kickUsersFromChannel', {
  type: 'Channel',
  args: { channelId: nonNull(stringArg()), userIds: nonNull(list(stringArg())) },
  description: 'Removes some users from [public] channel.',

  resolve: async (_, { channelId, userIds }, ctx) => {
    const userId = getUserId(ctx);
    const channel = await findExistingChannel(channelId);

    if (channel.channelType === 'private') {
      throw new Error('Removing users from the private channels is not allowed.');
    }

    const membership = await ctx.prisma.membership.findFirst({
      where: {
        userId,
        channelId: channel.id,
      },
    });

    const hasAdminPermission = membership.membershipType !== 'admin' && membership.membershipType !== 'owner';

    if (hasAdminPermission) {
      throw new Error('You should have admin or owner membership to add some users to the channel.');
    }

    await ctx.prisma.membership.deleteMany({
      where: {
        channelId,
        userId: {
          in: userIds,
        },
      },
    });

    return channel;
  },
});
