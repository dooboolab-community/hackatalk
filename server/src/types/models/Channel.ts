import { booleanArg, objectType } from '@nexus/schema';

import { getUserId } from '../../utils/auth';
import { relayToPrismaPagination } from '../../utils/pagination';

export const Channel = objectType({
  name: 'Channel',
  definition(t) {
    t.model.id();
    t.model.channelType();
    t.model.name();
    t.model.createdAt();
    t.model.updatedAt();
    t.model.deletedAt();

    t.field('lastMessage', {
      type: 'Message',
      nullable: true,
      description: 'Get latest message sent to the channel.',

      resolve: async ({ id }, args, ctx) => {
        const messages = await ctx.prisma.message.findMany({
          where: {
            channel: { id },
          },
          orderBy: { createdAt: 'desc' },
          take: 1,
        });

        return messages[0];
      },
    });

    t.connectionField('messages', {
      type: 'Message',

      nodes: async ({ id }, args, ctx) => {
        const { after, before, first, last } = args;

        return ctx.prisma.message.findMany({
          ...relayToPrismaPagination({
            after, before, first, last,
          }),
          where: { channel: { id } },
        });
      },
    });

    t.field('memberships', {
      type: 'Membership',
      list: true,
      nullable: true,
      description: 'Get memberships assigned to channel. If excludeMe is set, it will not return authenticated user.',
      args: { excludeMe: booleanArg() },

      resolve: ({ id }, { excludeMe }, ctx) => {
        const userId = getUserId(ctx);

        if (excludeMe) {
          return ctx.prisma.membership.findMany({
            where: {
              channel: { id },
              user: {
                id: { not: userId },
              },
            },
            include: { user: true },
          });
        }

        return ctx.prisma.membership.findMany({
          where: { channel: { id } },
          include: { user: true },
        });
      },
    });
  },
});
