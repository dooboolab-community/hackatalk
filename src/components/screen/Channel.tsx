import { ChannelType, MessageType } from '../../types';
import { FlatList, TouchableOpacity, View } from 'react-native';
import React, { useState } from 'react';

import ChannelListItem from '../shared/ChannelListItem';
import EmptyListItem from '../shared/EmptyListItem';
import { MainStackNavigationProps } from '../navigation/MainStackNavigator';
import { SvgPlus } from '../../utils/Icons';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useNavigation } from '@react-navigation/native';
import { useThemeContext } from '@dooboo-ui/native-theme';

const Container = styled.View`
  flex: 1;
  background: ${({ theme }): string => theme.background};
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Fab = styled.View`
  width: 56px;
  height: 56px;
  border-radius: 28px;
  justify-content: center;
  align-items: center;
  background: ${({ theme }): string => theme.fab};
`;

interface Props {
  navigation: MainStackNavigationProps<'Message'>;
}

const initialChannels: ChannelType[] = [
  {
    id: 'room1',
    lastMessage: {
      id: 'id_3',
      sender: {
        id: 'uid_3',
        nickname: 'displayName3',
        thumbURL: '',
        photoURL: '',
        statusMessage: '',
        isOnline: false,
      },
      messageType: MessageType.Message,
      message: 'How are you doing?',
      created: '2020-01-01 12:00',
      updated: '2020-01-01 12:00',
    },
    lastMessageCnt: 3,
  },
  {
    id: 'room2',
    lastMessage: {
      id: 'id_3',
      sender: {
        id: 'uid_3',
        nickname: 'Byun8585',
        thumbURL: '',
        photoURL: '',
        statusMessage: '',
        isOnline: false,
      },
      messageType: MessageType.Message,
      message: 'Hi. This is student from react-native-seoul. Nice to meet you.',
      created: '2020-01-01 12:00',
      updated: '2020-01-01 12:00',
    },
    lastMessageCnt: 0,
  },
];

function Channel(): React.ReactElement {
  const [channels] = useState(initialChannels);
  const { theme } = useThemeContext();
  const navigation = useNavigation();

  const onItemClick = (itemId: string): void => {
    navigation.navigate('Message', { messageId: itemId });
  };

  const renderItem = (item: ChannelType, index: number): React.ReactElement => {
    return (
      <ChannelListItem
        testID={`list-item-${index}`}
        item={item}
        onPress={(): void => onItemClick(item.id)}
      />
    );
  };
  return (
    <Container>
      <FlatList
        style={{
          alignSelf: 'stretch',
        }}
        contentContainerStyle={
          channels.length === 0
            ? {
              flex: 1,
              alignSelf: 'stretch',
              alignItems: 'center',
              justifyContent: 'center',
            }
            : null
        }
        keyExtractor={(_, index): string => index.toString()}
        data={channels}
        renderItem={({ item, index }): React.ReactElement =>
          renderItem(item, index)
        }
        ListEmptyComponent={
          <EmptyListItem>{getString('NO_CONTENT')}</EmptyListItem>
        }
      />
      <TouchableOpacity
        activeOpacity={0.65}
        style={{
          position: 'absolute',
          right: 20,
          bottom: 44,
        }}
        onPress={(): void => navigation.navigate('ChannelCreate')}
      >
        <Fab>
          <SvgPlus fill={theme.background}/>
        </Fab>
      </TouchableOpacity>
    </Container>
  );
}
export default Channel;
