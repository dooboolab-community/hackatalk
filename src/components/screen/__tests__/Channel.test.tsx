import 'react-native';

import * as React from 'react';

import { RenderResult, cleanup, render } from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import Channel from '../Channel';
import renderer from 'react-test-renderer';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: React.ReactElement;
let testingLib: RenderResult;

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: (): object => {
      return {
        navigation: {
          navigate: jest.fn(),
        },
      };
    },
  };
});

describe('[Channel] screen', () => {
  beforeEach(() => {
    props = createTestProps();
    component = createTestElement(<Channel {...props} />);
    testingLib = render(component);
  });

  it('renders without crashing', () => {
    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('interactions', () => {
    beforeEach(() => {
      testingLib = render(component);
    });

    it('should simulate onPress', () => {
      // const btn = testingLib.queryByTestId('btn');
      // act(() => {
      //   fireEvent.press(btn);
      //   fireEvent.press(btn);
      // });
      // expect(cnt).toBe(3);
    });
  });

  afterEach(() => {
    cleanup();
  });
});
