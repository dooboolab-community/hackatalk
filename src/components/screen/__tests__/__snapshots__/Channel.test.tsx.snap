// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Channel] screen renders without crashing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <ScrollView
        ListEmptyComponent={
          <Shared
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "flex": 1,
                "justifyContent": "center",
                "width": "100%",
              }
            }
          >
            No content
          </Shared>
        }
        contentContainerStyle={null}
        data={
          Array [
            Object {
              "id": "room1",
              "lastMessage": Object {
                "created": "2020-01-01 12:00",
                "id": "id_3",
                "message": "How are you doing?",
                "messageType": 0,
                "sender": Object {
                  "id": "uid_3",
                  "isOnline": false,
                  "nickname": "displayName3",
                  "photoURL": "",
                  "statusMessage": "",
                  "thumbURL": "",
                },
                "updated": "2020-01-01 12:00",
              },
              "lastMessageCnt": 3,
            },
            Object {
              "id": "room2",
              "lastMessage": Object {
                "created": "2020-01-01 12:00",
                "id": "id_3",
                "message": "Hi. This is student from react-native-seoul. Nice to meet you.",
                "messageType": 0,
                "sender": Object {
                  "id": "uid_3",
                  "isOnline": false,
                  "nickname": "Byun8585",
                  "photoURL": "",
                  "statusMessage": "",
                  "thumbURL": "",
                },
                "updated": "2020-01-01 12:00",
              },
              "lastMessageCnt": 0,
            },
          ]
        }
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        maxToRenderPerBatch={10}
        numColumns={1}
        onEndReachedThreshold={2}
        removeClippedSubviews={false}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "alignSelf": "stretch",
          }
        }
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View
          style={null}
        >
          <View
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <TouchableOpacity
              activeOpacity={0.5}
              testID="list-item-0"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#ffffff",
                      "borderBottomWidth": 1,
                      "borderColor": "rgb(229, 229, 229)",
                      "flexDirection": "row",
                      "height": 92,
                      "justifyContent": "space-between",
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "marginHorizontal": 20,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "width": 40,
                      }
                    }
                  >
                    <Text />
                  </View>
                  <View />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "flexBasis": 0,
                        "flexDirection": "column",
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "paddingRight": 20,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#151A25",
                            "fontSize": 14,
                            "fontWeight": "bold",
                          },
                        ]
                      }
                    >
                      displayName3
                    </Text>
                    <View
                      style={
                        Array [
                          Object {
                            "alignItems": "center",
                            "backgroundColor": "#00D4AB",
                            "borderRadius": 8,
                            "height": 16,
                            "justifyContent": "center",
                            "width": 16,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#1E6EFA",
                              "fontSize": 10,
                            },
                          ]
                        }
                      >
                        3
                      </Text>
                    </View>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                          "marginTop": 8,
                        },
                      ]
                    }
                  >
                    <Text
                      lastMessageCnt={3}
                      numberOfLines={2}
                      style={
                        Array [
                          Object {
                            "color": "#151A25",
                            "fontSize": 12,
                            "fontWeight": "bold",
                            "maxWidth": 150,
                          },
                        ]
                      }
                    >
                      How are you doing?
                    </Text>
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#1E6EFA",
                            "fontSize": 12,
                            "textAlign": "right",
                          },
                        ]
                      }
                    >
                      2 months ago
                    </Text>
                  </View>
                </View>
              </View>
            </TouchableOpacity>
          </View>
        </View>
        <View
          style={null}
        >
          <View
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <TouchableOpacity
              activeOpacity={0.5}
              testID="list-item-1"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "#ffffff",
                      "borderBottomWidth": 1,
                      "borderColor": "rgb(229, 229, 229)",
                      "flexDirection": "row",
                      "height": 92,
                      "justifyContent": "space-between",
                    },
                  ]
                }
              >
                <View
                  style={
                    Object {
                      "marginHorizontal": 20,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "height": 40,
                        "justifyContent": "center",
                        "width": 40,
                      }
                    }
                  >
                    <Text />
                  </View>
                  <View />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "flexBasis": 0,
                        "flexDirection": "column",
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "paddingRight": 20,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                        },
                      ]
                    }
                  >
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#151A25",
                            "fontSize": 14,
                            "fontWeight": "bold",
                          },
                        ]
                      }
                    >
                      Byun8585
                    </Text>
                    <View />
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "space-between",
                          "marginTop": 8,
                        },
                      ]
                    }
                  >
                    <Text
                      lastMessageCnt={0}
                      numberOfLines={2}
                      style={
                        Array [
                          Object {
                            "color": "#151A25",
                            "fontSize": 12,
                            "maxWidth": 150,
                          },
                        ]
                      }
                    >
                      Hi. This is student from react-native-seoul. Nice to meet you.
                    </Text>
                    <Text
                      style={
                        Array [
                          Object {
                            "color": "#1E6EFA",
                            "fontSize": 12,
                            "textAlign": "right",
                          },
                        ]
                      }
                    >
                      2 months ago
                    </Text>
                  </View>
                </View>
              </View>
            </TouchableOpacity>
          </View>
        </View>
      </ScrollView>
      <TouchableOpacity
        activeOpacity={0.65}
        style={
          Object {
            "bottom": 44,
            "position": "absolute",
            "right": 20,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#1E6EFA",
                "borderRadius": 28,
                "height": 56,
                "justifyContent": "center",
                "width": 56,
              },
            ]
          }
        >
          <SvgMock
            fill="#ffffff"
          />
        </View>
      </TouchableOpacity>
    </View>
  </View>
</View>
`;
